### ISSUE TITLE
Linear regression

### ISSUE THREAD

**Issue by vrxndx** [vrxndx/mllab#1]
#import all the libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import fetch_california_housing

#load the datset
housing = fetch_california_housing()
df = pd.DataFrame(housing.data, columns=housing.feature_names)

#add target variable to dataframes
df['MedHouseVal'] = housing.target
# display few rows
print(df.head())

#check missing values
print(df.isnull().sum())

# define X and y
X = df[['AveRooms']]
y = df['MedHouseVal'].values

#removing the outliers
q1, q3 = df['AveRooms'].quantile([0.25, 0.75])
iqr = q3 - q1
lower, upper = q1 - 1.5*iqr, q3 + 1.5*iqr

mask = (df['AveRooms'] >= lower) & (df['AveRooms'] <= upper)
df_clean = df.loc[mask].reset_index(drop=True)
print(f"Removed {len(df) - len(df_clean)} outlier rows; {len(df_clean)} remain.")
X = df_clean[['AveRooms']].values
y = df_clean['MedHouseVal'].values  # fixed: keep y 1D

# split (only once, after cleaning)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print(f"training set size: {X_train.shape}")
print(f"testing set size: {X_test.shape}")

#preprocessing after removing the outliers, not considering the total dataset
scalar = StandardScaler()
scalar.fit(X_train)
X_train_scaled = scalar.transform(X_train)
X_test_scaled = scalar.transform(X_test)
print("Mean/std used for scalling:", scalar.mean_, scalar.scale_)

# Normal Equation function
def normal_equation(X, y):
    # Add bias/intercept column
    X_b = np.c_[np.ones((X.shape[0], 1)), X]
    # Calculate theta = (XᵀX)^(-1)Xᵀy
    theta = np.linalg.inv(X_b.T @ X_b) @ X_b.T @ y
    return theta

# Train Normal Equation using SCALED training data
theta_ne = normal_equation(X_train_scaled, y_train)

# Extract bias and weight
b_ne = theta_ne[0]
w_ne = theta_ne[1]

print("Normal Equation Parameters:")
print("Weight (w):", w_ne)
print("Bias (b):", b_ne)

def predict(X, w, b):
    return w * X.ravel() + b

# fixed: predict on the SCALED test set, matching training scale
y_predict = predict(X_test_scaled, w_ne, b_ne)
print(y_predict)

x_line_scaled = np.linspace(X_test_scaled.min(), X_test_scaled.max(), 200).reshape(-1, 1)
y_line = predict(x_line_scaled, w_ne, b_ne)
x_line_original = scalar.inverse_transform(x_line_scaled).ravel()

plt.figure(figsize=(8, 6))
plt.scatter(X_test, y_test, alpha=0.3, s=15, label="test data")
plt.plot(x_line_original, y_line, color="red", linewidth=2.5, label="linear regression")
plt.xlabel("avg. rooms")
plt.ylabel("price")
plt.title("linear regression")
plt.legend()
plt.show()

def cost_compute(X, y, w, b):
    m = len(y)
    prediction = w * X.ravel() + b
    cost = (1/(2*m)) * np.sum((prediction - y)**2)
    return cost

def gradient(X, y, w_init=0.0, b_init=0.0, alpha=0.1, n_iteration=1000):
    X = X.reshape(-1)
    y = y.reshape(-1)
    m = len(y)
    w = float(w_init)
    b = float(b_init)
    cost_history = []

    for i in range(n_iteration):
        prediction = w * X + b
        error = prediction - y
        dw = np.sum(error * X) / m
        db = np.sum(error) / m
        w = w - alpha * dw
        b = b - alpha * db
        cost = np.sum(error**2) / (2*m)
        cost = float(cost)
        cost_history.append(cost)

        if i % 100 == 0:
            print(f"Iteration {i:4d}: Cost={cost:.5f}, w={w:.4f}, b={b:.4f}")

    return w, b, cost_history

w_gd, b_gd, cost_gd = gradient(
    X_train_scaled,
    y_train,
    w_init=0.0,
    b_init=0.0,
    alpha=0.1,
    n_iteration=1000
)

print(f"\nFinal parameters: w={w_gd:.4f}, b={b_gd:.4f}")

# --- Actual evaluation metrics (replaces old hardcoded comment block) ---
y_pred_ne = predict(X_test_scaled, w_ne, b_ne)
mse_ne = mean_squared_error(y_test, y_pred_ne)
r2_ne = r2_score(y_test, y_pred_ne)
print(f"\nNormal Equation -> MSE: {mse_ne:.3f}, R2: {r2_ne:.3f}")

y_pred_gd = predict(X_test_scaled, w_gd, b_gd)
mse_gd = mean_squared_error(y_test, y_pred_gd)
r2_gd = r2_score(y_test, y_pred_gd)
print(f"Gradient Descent -> MSE: {mse_gd:.3f}, R2: {r2_gd:.3f}")``
